src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php

```diff
@@ -69,15 +70,19 @@ class DoctrineChoiceLoader implements ChoiceLoaderInterface
      *                                                 IDs.
      * @param null|EntityLoaderInterface $objectLoader The objects loader
      */
-    public function __construct(ChoiceListFactoryInterface $factory, ObjectManager $manager, $class, IdReader $idReader = null, EntityLoaderInterface $objectLoader = null)
+    public function __construct(ChoiceListFactoryInterface $factory = null, ObjectManager $manager, $class, IdReader $idReader = null, EntityLoaderInterface $objectLoader = null)
     {
         $classMetadata = $manager->getClassMetadata($class);
 
-        $this->factory = $factory;
         $this->manager = $manager;
         $this->class = $classMetadata->getName();
         $this->idReader = $idReader ?: new IdReader($manager, $classMetadata);
         $this->objectLoader = $objectLoader;
+
+        if (null !== $this->factory) {
+            @trigger_error(sprintf('Passing a ChoiceListFactoryInterface to %s is deprecated since version 3.1 and will no longer be supported in 4.0. You should either call "%s::loadChoiceList" or override it to return a ChoiceListInterface.', __CLASS__, __CLASS__));
+            $this->factory = $factory;
+        }
```

Comment:
This deprecation means that in the future versions it will always have to be called with null as first argument. It's a pain to move the argument to the end of the list but I think that's the only solution you can do for 4.0